dcrpm/pidutil.py

```diff
@@ -14,16 +14,37 @@
 
 import psutil
 
-from dcrpm.util import TimeoutExpired, call_with_timeout
+from dcrpm.util import (
+    DcRPMException,
+    TimeoutExpired,
+    StatusCode,
+    call_with_timeout,
+    run_with_timeout,
+    which,
+)
 
 
 DEFAULT_TIMEOUT = 5  # seconds
+LSOF_TIMEOUT = 60  # seconds (OSX `lsof` is slow)
 MIN_PID = 2  # Don't kill init/launchd or kernel_task
 
 logger = logging.getLogger()
 
 
-def pids_holding_file(path):
+def process(pid):
+    # type: (int) -> Optional[psutil.Process]
+    """
+    Thin wrapper around psutil.Process with exception handling, mainly for
+    encapsulation.
+    """
+    try:
+        return psutil.Process(pid)
+    except psutil.NoSuchProcess:
+        logging.error("Pid %d does not exist or is no longer active", pid)
+        return None
+
+
+def procs_holding_file(path):
```

Comment:
This method can probably go away now, no?

Reply:
It could. The only reason to keep it is that `lsof` can be super slow (particularly on OS X). If that's not a worry I'll remove the psutil version.

Reply:
It may be worth to keep this as a fallback in case lsof times out

Reply:
I mean either we should make it an option to use the psutil version, or make it a fallback, or remove it... at the moment it's unreachable code...

Reply:
It's not unreachable — the lock file checks use `procs_holding_file` still.

I vote for just switching wholesale to `lsof`. Having a fallback is complicated for two reasons:

1) The point of the `lsof` code is to find holders of the DB by using the new link name. This approach doesn't work with the old psutil-based code. We'd have to find processes that opened `__db.001`, not the new link, when using the old code.

2) This PR makes the recovery happen *before* killing processes to avoid races. That in turn means finding processes that opened `__db.001` still won't work, because by then that file has been deleted. On Linux we could conceivably look for `__db.001 (deleted)` but that's not portable.

Reply:
I'm fine with replacing the process-searching code with lsof. Doing this properly in a cross-platform way is difficult, which is why I originally opted for psutil in the first place. Now that we have the extra requirement of ensuring hard links are found, it makes more sense to remove the psutil routines and only use lsof instead.

Unfortunately, as pointed out, lsof perf sucks on macOS > 10.13 currently :/

Reply:
My instinct is the same as @markdrayton 's.... but I guess the question is *how* bad is that perf, @get9 ? Are we talking minutes, or seconds? If it's bad enough we can avoid this whole check on mac, but if we're gonna do it, we should do it right, which requires `lsof`.

Reply:
5s on lightly loaded Mac server, 26s on my Macbook Pro.

Reply:
Woof. Holy hell that's bad. But... the alternative is not fixing macs... so unless there's some more-performant way, it feels like we should just bite the bullet...